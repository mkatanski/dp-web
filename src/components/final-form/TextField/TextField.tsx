import React from "react";
import { FieldProps, FieldRenderProps, useField } from "react-final-form";
import { TextField as MuiTextField, TextFieldProps } from "@material-ui/core";
import styled from "styled-components";

export type TextInputProps = { testId?: string; id?: string } & FieldProps<
  {},
  FieldRenderProps<TextFieldProps>
>;

export const TextFieldElement = styled(MuiTextField)`
  .MuiFormLabel-root.Mui-error {
    > span {
      color: ${({ theme }) => theme.textSecondary};
      opacity: 0.6;
    }
    & .MuiOutlinedInput-notchedOutline {
      border-color: ${({ theme }) => theme.textSecondary};
    }
  }

  .MuiFormHelperText-root {
    padding-top: 5px;
    font-size: 14px;
  }

  .MuiSvgIcon-root {
    color: ${({ theme }) => theme.textSecondary};
  }

  .MuiInput-underline {
    &:hover:not(.Mui-disabled):before {
      border-bottom: 1px solid ${({ theme }) => theme.palette.primary.light};
    }

    &:before,
    &:after {
      border-bottom: 1px solid ${({ theme }) => theme.palette.primary.light};
    }
  }

  && {
    margin: 10px 0;
  }

  fieldset {
    outline: none;
  }
`;

export const TextField: React.FC<TextInputProps> = ({
  name,
  testId = `rff-textfield-${name}`,
  id = name,
  fullWidth = true,
  variant = "outlined",
  ...props
}: TextInputProps) => {
  const { input, meta } = useField(name, props);

  return (
    <TextFieldElement
      {...input}
      id={id}
      label={props.label}
      error={meta.touched && !!meta.error}
      inputProps={{
        "data-testid": testId
      }}
      fullWidth={fullWidth}
      variant={variant}
      helperText={meta.touched && !!meta.error && meta.error}
      multiline={props.multiline}
      rows={props.rows}
    />
  );
};
