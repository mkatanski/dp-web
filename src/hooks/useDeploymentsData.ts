import { useSelector } from "react-redux";
import _ from "lodash";

import { RootReducerType } from "store";
import { DeploymentState } from "store/deployments";

export const useDeploymentsData = () => {
  const deploymentsStore = useSelector<RootReducerType, DeploymentState>(
    state => state.deploymentsReducer,
    (left, right) => _.isEqual(left.deployments, right.deployments)
  );

  return deploymentsStore.deployments;
};
