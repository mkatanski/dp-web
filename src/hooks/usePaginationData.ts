import { useSelector } from "react-redux";
import _ from "lodash";

import { RootReducerType, ReducerName } from "store";

type GenericListState = { limit: number; offset: number; totalCount: number };

export const usePaginationData = (name: ReducerName) => {
  const deploymentsStore = useSelector<RootReducerType, GenericListState>(
    state => state[name] as GenericListState,
    (left, right) => {
      const leftPagination = {
        limit: left.limit,
        offset: left.offset,
        total: left.totalCount
      };

      const rightPagination = {
        limit: right.limit,
        offset: right.offset,
        total: right.totalCount
      };

      return _.isEqual(leftPagination, rightPagination);
    }
  );

  const { limit, offset, totalCount } = deploymentsStore;
  return { limit, offset, totalCount };
};
