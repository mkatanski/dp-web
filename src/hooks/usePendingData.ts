import { useSelector } from "react-redux";
import _ from "lodash";

import { RootReducerType } from "store";
import { DeploymentState } from "store/deployments";

export const usePendingData = () => {
  const deploymentsStore = useSelector<RootReducerType, DeploymentState>(
    state => state.deploymentsReducer,
    (left, right) => _.isEqual(left.pendingList, right.pendingList)
  );

  return deploymentsStore.pendingList;
};
