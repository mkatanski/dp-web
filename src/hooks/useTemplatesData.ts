import { useSelector } from "react-redux";
import _ from "lodash";

import { RootReducerType } from "store";
import { TemplateState } from "store/templates";

export const useTemplatesData = () => {
  const templatesStore = useSelector<RootReducerType, TemplateState>(
    state => state.templatesReducer,
    (left, right) => _.isEqual(left.templates, right.templates)
  );

  return templatesStore.templates;
};
