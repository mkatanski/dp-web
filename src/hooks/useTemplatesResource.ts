import axios from "axios";
import env from "config/vars";

import { useDispatch, useStore } from "react-redux";
import { RootReducerType } from "store";
import { setTemplates } from "store/templates";

export type TemplateItem = {
  name: string;
  versions: string[];
  _id: string;
};

export const useTemplatesResource = () => {
  const dispatch = useDispatch();
  const { getState } = useStore<RootReducerType>();

  const fetchData = async (overrides?: { offset?: number; limit?: number }) => {
    const {
      offset: storeOffset,
      limit: storeLimit
    } = getState().templatesReducer;

    const offset = Number.isInteger(overrides?.offset)
      ? overrides?.offset
      : storeOffset;
    const limit = Number.isInteger(overrides?.limit)
      ? overrides?.limit
      : storeLimit;

    try {
      const result = await axios.get(
        `${env.FULL_PUBLIC_API_URL}/templates?limit=${limit}&offset=${offset}`
      );
      if (result.status >= 300 && result.status < 200) {
        throw new Error(result.statusText);
      }

      const items = result.data.items as TemplateItem[];

      dispatch(
        setTemplates(
          items.map(item => ({
            _id: item._id,
            name: item.name,
            versions: item.versions
          })),
          result.data.limit,
          result.data.offset,
          result.data.totalCount
        )
      );
    } catch (e) {
      // Do nothing for now
    }
  };

  return { fetchData };
};
