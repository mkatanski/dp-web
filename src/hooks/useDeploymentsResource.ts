import axios from "axios";
import env from "config/vars";

import { useDispatch } from "react-redux";
import { setDeployments } from "store/deployments";
import { usePaginationData } from "./usePaginationData";

export type DeploymentItem = {
  deployedAt: string;
  templateName: string;
  url: string;
  version: string;
  _id: string;
};

export type DeploymentPostBody = {
  url: string;
  templateName: string;
  version: string;
};

export const useDeploymentsResource = () => {
  const dispatch = useDispatch();
  const { limit, offset } = usePaginationData("deploymentsReducer");

  const fetchData = async () => {
    try {
      const result = await axios.get(
        `${env.FULL_PUBLIC_API_URL}/deployments?limit=${limit}&offset=${offset}`
      );
      if (result.status >= 300 && result.status < 200) {
        throw new Error(result.statusText);
      }

      const items = result.data.items as DeploymentItem[];

      dispatch(
        setDeployments(
          items.map(item => ({
            _id: item._id,
            pending: false,
            templateName: item.templateName,
            version: item.version
          })),
          result.data.limit,
          result.data.offset,
          result.data.totalCount
        )
      );
    } catch (e) {
      // Do nothing for now
    }
  };

  const postData = async (body: DeploymentPostBody) => {
    try {
      const result = await axios.post(
        `${env.FULL_PUBLIC_API_URL}/deployments`,
        body
      );
      if (result.status >= 300 && result.status < 200) {
        throw new Error(result.statusText);
      }

      const item = result.data.item as DeploymentItem[];
      return item;
    } catch (e) {
      // Do nothing for now
    }
  };

  return { fetchData, postData };
};
