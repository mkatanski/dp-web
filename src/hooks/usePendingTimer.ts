import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { pendingTimerTick } from "store/deployments";
import { usePendingData } from "./usePendingData";

export const usePendingTimer = () => {
  const dispatch = useDispatch();
  const pending = usePendingData();
  const pendingCount = Object.keys(pending).length;

  useEffect(() => {
    let iHandle: number;
    if (pendingCount) {
      iHandle = setInterval(() => {
        dispatch(pendingTimerTick());
      }, 1000);
    }

    return () => {
      iHandle && clearInterval(iHandle);
    };
  }, [dispatch, pendingCount]);
};
