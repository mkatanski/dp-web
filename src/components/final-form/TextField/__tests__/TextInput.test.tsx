/* eslint-disable react/display-name */
import React from "react";
import TestRenderer from "react-test-renderer";
import "jest-styled-components";
import { useField } from "hooks/useField";

import { TextField } from "../TextField";

// Uncomment below if You need/want to use enzyme
// import enzyme from "enzyme";

jest.mock("components/forms/TextField/TextFieldElement", () => ({
  TextFieldElement: ({ ...rest }: React.FC<{}>) => (
    <input id="TextFieldElement" data-prop={rest} />
  )
}));

jest.mock("hooks/useField", () => ({
  useField: jest.fn()
}));

const mockUseField = useField as jest.Mock;

describe("components", () => {
  describe("final-form", () => {
    describe("TextField", () => {
      it("should handle case when field is valid", () => {
        mockUseField.mockReturnValue({
          input: {},
          meta: {
            touched: true,
            error: false
          }
        });
        const validateFn = jest.fn();
        const tree = TestRenderer.create(
          <TextField name="TextField" validate={validateFn} />
        );
        expect(mockUseField).toHaveBeenCalledWith("TextField", {
          type: undefined,
          validate: validateFn
        });
        expect(tree).toMatchSnapshot();
      });

      it("should handle case when there is an error", () => {
        mockUseField.mockReturnValue({
          input: {},
          meta: {
            touched: true,
            error: "This field is invalid"
          }
        });
        const validateFn = jest.fn();
        const tree = TestRenderer.create(
          <TextField name="TextField" validate={validateFn} />
        );
        expect(mockUseField).toHaveBeenCalledWith("TextField", {
          type: undefined,
          validate: validateFn
        });
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
