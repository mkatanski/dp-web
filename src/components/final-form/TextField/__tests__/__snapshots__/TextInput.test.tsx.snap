// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components final-form TextField should handle case when field is valid 1`] = `
<input
  data-prop={
    Object {
      "error": false,
      "fullWidth": true,
      "helperText": false,
      "id": "TextField",
      "inputProps": Object {
        "data-testid": "rff-textfield-TextField",
      },
      "label": undefined,
      "multiline": undefined,
      "rows": undefined,
      "variant": "outlined",
    }
  }
  id="TextFieldElement"
/>
`;

exports[`components final-form TextField should handle case when there is an error 1`] = `
<input
  data-prop={
    Object {
      "error": true,
      "fullWidth": true,
      "helperText": "This field is invalid",
      "id": "TextField",
      "inputProps": Object {
        "data-testid": "rff-textfield-TextField",
      },
      "label": undefined,
      "multiline": undefined,
      "rows": undefined,
      "variant": "outlined",
    }
  }
  id="TextFieldElement"
/>
`;
